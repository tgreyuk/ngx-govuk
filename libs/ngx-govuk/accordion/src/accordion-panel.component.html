<p>accordion-panel works!</p>
