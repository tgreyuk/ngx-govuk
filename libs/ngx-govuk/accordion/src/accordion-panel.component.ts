import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'ngx-govuk-accordion-panel',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './accordion-panel.component.html',
  styleUrl: './accordion-panel.component.scss',
})
export class AccordionPanelComponent {}
