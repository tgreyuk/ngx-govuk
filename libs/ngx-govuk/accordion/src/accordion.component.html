<div class="govuk-accordion">
  <div class="govuk-accordion__controls">
    <button
      type="button"
      class="govuk-accordion__show-all"
      aria-expanded="false"
      (click)="toggle()"
    >
      <span
        class="govuk-accordion-nav__chevron"
        [ngClass]="{
          'govuk-accordion-nav__chevron--down': !isExpanded()
        }"
      ></span
      ><span class="govuk-accordion__show-all-text">{{
        isExpanded() ? 'Hide all sections' : 'Show all sections'
      }}</span>
    </button>
  </div>
  <ng-content></ng-content>
</div>
