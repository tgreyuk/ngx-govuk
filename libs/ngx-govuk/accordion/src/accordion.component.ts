import { CommonModule } from '@angular/common';
import { Component } from '@angular/core';

@Component({
  selector: 'ngx-govuk-accordion',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './accordion.component.html',
  styleUrl: './accordion.component.scss',
})
export class NgxGovukAccordionComponent {}
