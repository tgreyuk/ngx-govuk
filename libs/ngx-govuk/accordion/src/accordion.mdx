import { Meta, Story, ArgsTable, Canvas } from '@storybook/blocks';
import * as stories from './accordion.component.stories';

<Meta of={stories} />

# Accordion

The accordion component lets users show and hide sections of related content on a page.

## GOV.UK Docs

[https://design-system.service.gov.uk/components/accordion/](https://design-system.service.gov.uk/components/accordion/)

## Usage

```ts
import { NgxGovukAccordionComponent } from 'ngx-govuk/accordion';

@Component({
  standalone: true,
  imports: [NgxGovukAccordionComponent],
  selector: 'my-component',
  template: `<ngx-govuk-accordion></ngx-govuk-accordion>`,
})
```
