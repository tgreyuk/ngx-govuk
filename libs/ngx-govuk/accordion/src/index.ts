export * from './accordion-panel.component';
export * from './accordion.component';
