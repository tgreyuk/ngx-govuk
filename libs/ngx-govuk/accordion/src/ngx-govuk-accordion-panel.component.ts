import { CommonModule } from '@angular/common';
import {
  ChangeDetectionStrategy,
  Component,
  Input,
  signal,
} from '@angular/core';
import { nanoid } from 'nanoid';

@Component({
  selector: 'ngx-govuk-accordion-panel',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './ngx-govuk-accordion-panel.component.html',
  styleUrl: './ngx-govuk-accordion-panel.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class NgxGovukAccordionPanelComponent {
  @Input({ required: true }) heading!: string;

  isExpanded = signal(false);

  headingId = nanoid();
  contentId = nanoid();

  toggle() {
    this.isExpanded.set(!this.isExpanded());
  }
}
