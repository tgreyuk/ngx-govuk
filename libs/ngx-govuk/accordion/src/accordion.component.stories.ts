import { moduleMetadata, type Meta, type StoryObj } from '@storybook/angular';
import { NgxGovukAccordionPanelComponent } from './accordion-panel.component';
import { NgxGovukAccordionComponent } from './accordion.component';

const meta: Meta<NgxGovukAccordionComponent> = {
  component: NgxGovukAccordionComponent,
  title: 'Accordion',
  decorators: [
    moduleMetadata({
      imports: [NgxGovukAccordionComponent, NgxGovukAccordionPanelComponent],
    }),
  ],
};
export default meta;
type Story = StoryObj<NgxGovukAccordionComponent>;

export const Preview: Story = {
  args: {},
  render: (args) => ({
    props: args,
    template: `
<div class="govuk-frontend-supported">
  <ngx-govuk-accordion>
    <ngx-govuk-accordion-panel heading="Writing well for the web"><p class="govuk-body">This is the content for Writing well for the web.</p></ngx-govuk-accordion-panel>
    <ngx-govuk-accordion-panel heading="Writing well for specialists"><p class="govuk-body">This is the content for Writing well for specialists.</p></ngx-govuk-accordion-panel>
    <ngx-govuk-accordion-panel heading="Know your audience"><p class="govuk-body">This is the content for Know your audience.</p></ngx-govuk-accordion-panel>
    <ngx-govuk-accordion-panel heading="How people read"><p class="govuk-body">This is the content for How people read.</p></ngx-govuk-accordion-panel>
  </ngx-govuk-accordion>
</div>
`,
  }),
};
