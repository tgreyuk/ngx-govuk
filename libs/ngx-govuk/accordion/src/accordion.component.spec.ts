import { Component, DebugElement } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { MockProvider } from 'ng-mocks';
import { NgxGovukUtilsService } from '../../src';
import { NgxGovukAccordionPanel } from './accordion-panel.component';
import { NgxGovukAccordion } from './accordion.component';

@Component({
  standalone: true,
  imports: [NgxGovukAccordion, NgxGovukAccordionPanel],
  providers: [
    MockProvider(NgxGovukUtilsService, {
      randomId(prefix: string): string {
        return prefix + '-someId';
      },
    }),
  ],
  template: `<ngx-govuk-accordion>
    <ngx-govuk-accordion-panel
      heading="Test heading"
    ></ngx-govuk-accordion-panel>
  </ngx-govuk-accordion>`,
})
class TestHostComponent {}

describe('NgxGovukAccordion', () => {
  let fixture: ComponentFixture<TestHostComponent>;
  let component: NgxGovukAccordion;
  let debugElement: DebugElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [TestHostComponent, NgxGovukAccordion, NgxGovukAccordionPanel],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    fixture.detectChanges();
    debugElement = fixture.debugElement.query(By.directive(NgxGovukAccordion));
    component = debugElement.componentInstance;
    component.isExpanded.set(false);
  });

  it('should create accordion component', () => {
    expect(component).toBeTruthy();
  });

  it('should render default state', () => {
    expect(debugElement.nativeNode).toMatchSnapshot();
  });

  it('should render expanded state', () => {
    component.toggle();
    fixture.detectChanges();
    expect(debugElement.nativeNode).toMatchSnapshot();
  });
});
