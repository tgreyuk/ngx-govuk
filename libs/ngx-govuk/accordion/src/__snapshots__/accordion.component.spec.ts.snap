// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgxGovukAccordion should render default state 1`] = `
<ngx-govuk-accordion>
  <div
    class="govuk-accordion"
  >
    <div
      class="govuk-accordion__controls"
    >
      <button
        aria-expanded="false"
        class="govuk-accordion__show-all"
        type="button"
      >
        <span
          class="govuk-accordion-nav__chevron govuk-accordion-nav__chevron--down"
        />
        <span
          class="govuk-accordion__show-all-text"
        >
          Show all sections
        </span>
      </button>
    </div>
    <ngx-govuk-accordion-panel
      class="govuk-accordion__section"
      heading="Test heading"
    >
      <div
        class="govuk-accordion__section-header"
      >
        <h2
          class="govuk-accordion__section-heading"
        >
          <button
            aria-controls="content-someId"
            aria-expanded="false"
            aria-label="Hide this section"
            class="govuk-accordion__section-button"
            type="button"
          >
            <span
              class="govuk-accordion__section-heading-text"
              id="heading-someId"
            >
              <span
                class="govuk-accordion__section-heading-text-focus"
              >
                 Test heading 
              </span>
            </span>
            <span
              class="govuk-visually-hidden govuk-accordion__section-heading-divider"
            >
              , 
            </span>
            <span
              class="govuk-accordion__section-toggle"
            >
              <span
                class="govuk-accordion__section-toggle-focus"
              >
                <span
                  class="govuk-accordion-nav__chevron govuk-accordion-nav__chevron--down"
                />
                <span
                  class="govuk-accordion__section-toggle-text"
                >
                  Show
                </span>
              </span>
            </span>
          </button>
        </h2>
      </div>
      <div
        aria-labelledby="heading-someId"
        class="govuk-accordion__section-content"
        id="content-someId"
      />
    </ngx-govuk-accordion-panel>
  </div>
</ngx-govuk-accordion>
`;

exports[`NgxGovukAccordion should render expanded state 1`] = `
<ngx-govuk-accordion>
  <div
    class="govuk-accordion"
  >
    <div
      class="govuk-accordion__controls"
    >
      <button
        aria-expanded="true"
        class="govuk-accordion__show-all"
        type="button"
      >
        <span
          class="govuk-accordion-nav__chevron"
        />
        <span
          class="govuk-accordion__show-all-text"
        >
          Hide all sections
        </span>
      </button>
    </div>
    <ngx-govuk-accordion-panel
      class="govuk-accordion__section govuk-accordion__section--expanded"
      heading="Test heading"
    >
      <div
        class="govuk-accordion__section-header"
      >
        <h2
          class="govuk-accordion__section-heading"
        >
          <button
            aria-controls="content-someId"
            aria-expanded="true"
            aria-label="Hide this section"
            class="govuk-accordion__section-button"
            type="button"
          >
            <span
              class="govuk-accordion__section-heading-text"
              id="heading-someId"
            >
              <span
                class="govuk-accordion__section-heading-text-focus"
              >
                 Test heading 
              </span>
            </span>
            <span
              class="govuk-visually-hidden govuk-accordion__section-heading-divider"
            >
              , 
            </span>
            <span
              class="govuk-accordion__section-toggle"
            >
              <span
                class="govuk-accordion__section-toggle-focus"
              >
                <span
                  class="govuk-accordion-nav__chevron"
                />
                <span
                  class="govuk-accordion__section-toggle-text"
                >
                  Hide
                </span>
              </span>
            </span>
          </button>
        </h2>
      </div>
      <div
        aria-labelledby="heading-someId"
        class="govuk-accordion__section-content"
        id="content-someId"
      />
    </ngx-govuk-accordion-panel>
  </div>
</ngx-govuk-accordion>
`;
